packages/gatsby-source-contentful/src/source-nodes.js

```diff
@@ -0,0 +1,221 @@
+ * 3. When creating entries and assets, make the most localized version
+ * possible for each localized node i.e. get the localized field if it exists
+ * or the fallback field or the default field.
+ */
+
+export async function sourceNodes(
+  {
+    actions,
+    getNode,
+    getNodes,
+    getNodesByType,
+    createNodeId,
+    store,
+    cache,
+    getCache,
+    reporter,
+    parentSpan,
+  },
+  pluginOptions
+) {
+  const { createNode, touchNode } = actions
+
+  const pluginConfig = createPluginConfig(pluginOptions)
+  const sourceId = `${pluginConfig.get(`spaceId`)}-${pluginConfig.get(
+    `environment`
+  )}`
+
+  const {
+    mergedSyncData,
+    contentTypeItems,
+    locales,
+    space,
+    defaultLocale,
+  } = await cache.get(`contentful-sync-result-${sourceId}`)
+
+  const { assets } = mergedSyncData
+
+  const entryList = normalize.buildEntryList({
+    mergedSyncData,
+    contentTypeItems,
+  })
+
+  const existingNodes = getNodes().filter(
+    n => n.internal.owner === `gatsby-source-contentful`
+  )
+  existingNodes.forEach(n => touchNode(n))
+
+  reporter.verbose(`Building Contentful reference map`)
+
+  // Create map of resolvable ids so we can check links against them while creating
+  // links.
+  const resolvable = normalize.buildResolvableSet({
+    existingNodes,
+    entryList,
+    assets,
+    defaultLocale,
+    locales,
+    space,
+  })
+
+  // Build foreign reference map before starting to insert any nodes
+  const foreignReferenceMap = normalize.buildForeignReferenceMap({
+    contentTypeItems,
+    entryList,
+    resolvable,
+    defaultLocale,
+    locales,
+    space,
+    useNameForId: pluginConfig.get(`useNameForId`),
+  })
+
+  reporter.verbose(`Resolving Contentful references`)
+
+  const newOrUpdatedEntries = new Set()
+  entryList.forEach(entries => {
+    entries.forEach(entry => {
+      newOrUpdatedEntries.add(`${entry.sys.id}___${entry.sys.type}`)
+    })
+  })
+
+  // Update existing entry nodes that weren't updated but that need reverse
+  // links added.
+  existingNodes
+    .filter(n => newOrUpdatedEntries.has(`${n.id}___${n.sys.type}`))
+    .forEach(n => {
+      if (foreignReferenceMap[`${n.id}___${n.sys.type}`]) {
+        foreignReferenceMap[`${n.id}___${n.sys.type}`].forEach(
+          foreignReference => {
+            // Add reverse links
+            if (n[foreignReference.name]) {
+              n[foreignReference.name].push(foreignReference.id)
+              // It might already be there so we'll uniquify after pushing.
+              n[foreignReference.name] = _.uniq(n[foreignReference.name])
+            } else {
+              // If is one foreign reference, there can always be many.
+              // Best to be safe and put it in an array to start with.
+              n[foreignReference.name] = [foreignReference.id]
+            }
+          }
+        )
+      }
+    })
+
+  const creationActivity = reporter.activityTimer(
```

Comment:
@todo consider including all the map creation and loops in the creation activity.